chandra_aca/aca_image.py

```diff
@@ -309,6 +311,257 @@ def col0(self):
+
+        # Make a bool ACAImage like self to allow convenient mask/unmask of
+        # pixels to flicker.  This is used in annie.  Also make the corresponding
+        # 1-d ravelled version.
+        self.flicker_mask = ACAImage(np.ones(self.shape, dtype=np.bool),
+                                     row0=self.row0, col0=self.col0)
+        self.flicker_mask_vals = self.flicker_mask.view(np.ndarray).ravel()
+
+        # Get the index to the CDFs which is appropriate for each pixel
+        # based on its initial value.
+        self.flicker_cdf_idxs = np.searchsorted(self.flicker_cdf_bins,
+                                                self.flicker_vals0) - 1
+
+        # Create an array of time (secs) until next flicker for each pixel
+        if seed == -1:
+            # Special case of testing, use a constant flicker_mean_time initially
+            t_flicker = np.ones(shape=(self.flicker_n_vals,)) * flicker_mean_time
+            phase = 1.0
+        else:
+            # This is drawing from an exponential distribution.  For the initial
+            # time assume the flickering is randomly phased within that interval.
+            phase = np.random.uniform(0.0, 1.0, size=self.flicker_n_vals)
+            rand_unifs = np.random.uniform(0.0, 1.0, size=self.flicker_n_vals)
+            t_flicker = -np.log(1.0 - rand_unifs) * flicker_mean_time
+
+        self.flicker_times = t_flicker * phase
+
+    def flicker_update(self, dt, use_numba=True):
+        """Propagate the image forward by ``dt`` seconds and update any pixels
+        that have flickered during that interval.
+
+        This has the option to use one of two implementations.  The default is
+        to use the numba-based version which is about 6 times faster.  The
+        vectorized version is left in for reference.
+
+        :param dt: time (secs) to propagate image
+        :param use_numba: use the numba version of updating (default=True)
+        """
+        if not hasattr(self, 'flicker_times'):
+            self.flicker_init()
+
+        if use_numba:
+            _flicker_update_numba(dt,
+                                  len(self.flicker_vals),
+                                  self.test_idx,
+                                  self.flicker_vals0,
+                                  self.flicker_vals,
+                                  self.flicker_mask_vals,
+                                  self.flicker_times,
+                                  self.flicker_cdf_idxs,
+                                  self.flicker_cdf_x,
+                                  self.flicker_cdfs,
+                                  self.flicker_scale,
+                                  self.flicker_mean_time)
+            if self.test_idx > 0:
+                self.test_idx += 1
+        else:
+            self._flicker_update_vectorized(dt)
+
+    def _flicker_update_vectorized(self, dt):
+
+        self.flicker_times[self.flicker_mask_vals] -= dt
```

Comment:
Should this also not bother to update the flicker times for the pixels with the flicker_cdf_idxs < 0 ?

Reply:
Ah, but I suppose you don't care about the vectorized implementation anyway so moot.